File: NeetCode/Math_And_Geometry/Non_Cyclical_Number/non__cyclical__number.py
from typing import Optional, List


class Solution:
    def isHappy(self, n: int) -> bool:
        def calc(i):
            s = str(i)
            sum = 0
            for ch in s:
                sum += int(ch) ** 2
            return sum

        seen = set()

        while n != 1:
            if n in seen:
                return False
            seen.add(n)
            n = calc(n)
        return True


if __name__ == "__main__":
    sol = Solution()
    print(sol.isHappy(100))
    print(sol.isHappy(101))
    print("Running Solution...")

File: NeetCode/Math_And_Geometry/Non_Cyclical_Number/Non_Cyclical_Number.cpp
// Non_Cyclical_Number.cpp
#include <algorithm>
#include <array>
#include <cassert>
#include <cctype>
#include <climits>
#include <cmath>
#include <deque>
#include <functional>
#include <iostream>
#include <map>
#include <numeric>
#include <queue>
#include <set>
#include <stack>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <vector>

using namespace std;

class Solution
{
public:
    bool isHappy(int n)
    {
        unordered_set<int> seen;
        while (n != 1)
        {
            if (seen.count(n))
            {
                return false;
            }
            seen.insert(n);
            n = sqSum(n);
        }
        return true;
    }
    int sqSum(int n)
    {
        int total = 0;
        while (n)
        {
            total = total + (n % 10) * (n % 10);
            n = n / 10;
        }
        return total;
    }
};
